File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
        // if(n == 0) return 1.0;
        // if(x == 0) return 0.0;
        // if(x == 1) return 1.0;
        // if(x == -1 && n % 2 == 0) return 1.0;
        // if(x == -1 && n % 2 == 0) return -1.0;

//         long binForm = n;
//         //if n is negative then we have to convert 
//         if(n < 0){                      
//             x = 1/x;          
//             binForm = -binForm;//binform is negative it will convert to positive
//         }
//         int ans = 1;

//         while(binForm > 0){
//             if(binForm % 2 == 1){
//                 ans *= x;
//             }
//             x *=x;
//             binForm /=2;
//         }
//     return ans;
//     }
// };




        long binForm = n;
        if(n < 0){
            x = 1/x;
            binForm = -binForm;
        }

        double ans = 1;

        while(binForm > 0){
            if(binForm % 2 == 1){
                ans *= x;
            }
            x *= x;
            binForm /= 2;
        }
        return ans;
    }
};

File: 0050-powx-n/0050-powx-n.java
class Solution {
    double solve(double x, long n){
        if(n == 0) return 1.0;
        if(n < 0){
            return solve(1/x, -n);
        }
        if(n % 2 == 0){
            return solve(x*x, n/2);
        }
        else return x*solve(x*x, (n-1)/2);
    }

    public double myPow(double x, long n) {
        return solve(x, (long)n);
    }   
}

// class Solution {
//     double solve(double x, long n) {
//         if (n == 0) return 1.0;
        
//         if (n % 2 == 0) {
//             return solve(x * x, n / 2);
//         } else {
//             return x * solve(x * x, (n - 1) / 2);
//         }
//     }

//     public double myPow(double x, int n) {
//         long power = n;  // Convert to long to avoid overflow
//         if (power < 0) {
//             x = 1 / x;
//             power = -power;
//         }
//         return solve(x, power);
//     }
// }
